### PR TITLE
feat(inverted_index): inverted index cache

### PR DESCRIPTION
**Pull request by v0y4g3r**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

This PR adds im-memory cache for inverted index that boost most queries by reducing the pruning cost, which reduces total cost by 2~3ms:
image
 
## Future work

- [ ] Cache for Puffin file metadata is still missing and I will not include that in this PR since `PuffinReader` has been refactored recently and I need to find a proper way to pass `InvertedIndexCache` to `BoundedStager`.

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

## Summary by CodeRabbit

- **New Features**
  - Introduced caching functionality for inverted index data to improve performance.
  - Added configuration settings for managing cache sizes of inverted index metadata and content.

- **Improvements**
  - Enhanced the `SstIndexApplier` to conditionally use the cache for index reading.
  - Modified methods to return shared ownership of metadata using `Arc`.
  
- **Configuration Updates**
  - Added settings for inverted index metadata and content cache sizes in configuration files.

- **Testing Enhancements**
  - Updated test cases to include new parameters for index cache.

### CODE REVIEW

**Comment by killme2008** [comment]
Do we have any benchmark program and results for this PR? @v0y4g3r

**Comment by v0y4g3r** [comment]
> Do we have any benchmark program and results for this PR? @v0y4g3r

I updated the PR description. This PR improves those queries which are already fast, but not fast enough, namely those cost about 5~6ms. An improvement of 2~3ms means cutting the total cost time by half.

File: src/mito2/src/config.rs [hunk 1]
```diff
@@ -84,6 +84,10 @@ pub struct MitoConfig {
     pub vector_cache_size: ReadableSize,
     /// Cache size for pages of SST row groups. Setting it to 0 to disable the cache.
     pub page_cache_size: ReadableSize,
+    /// Cache size for metadata of inverted index. Setting it to 0 to disable the cache.
+    pub inverted_index_metadata_cache_size: ReadableSize,
+    /// Cache size for inverted index content. Setting it to 0 to disable the cache.
+    pub inverted_index_cache_size: ReadableSize,
```
Comment (zhongzc) [inline src/mito2/src/config.rs:90]: How about putting them in `InvertedIndexConfig`?

**Review by zhongzc — APPROVED** [review]
most LGTM

File: src/mito2/src/cache/index.rs [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+            let key = IndexKey {
+                file_id: self.file_id,
+            };
+            let value = Arc::new(all_data);
+            let size = index_content_weight(&key, &value);
+            self.cache.put_index(key, value);
+            CACHE_MISS.with_label_values(&[INDEX_CONTENT_TYPE]).inc();
+            CACHE_BYTES
+                .with_label_values(&[INDEX_CONTENT_TYPE])
+                .add(size.into());
```
Comment (evenyag) [inline src/mito2/src/cache/index.rs:81]: Looks like `put_index()` already added `CACHE_BYTES`.

File: src/mito2/src/cache/index.rs [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+use crate::metrics::{CACHE_BYTES, CACHE_HIT, CACHE_MISS};
+
+/// Metrics for index metadata.
+const INDEX_METADATA_TYPE: &str = "index_metadata";
+/// Metrics for index content.
+const INDEX_CONTENT_TYPE: &str = "index_content";
+
+/// Inverted index blob reader with cache.
+pub struct CachedInvertedIndexBlobReader<R> {
+    file_id: Uuid,
```
Comment (evenyag) [inline src/mito2/src/cache/index.rs:36]: Can we use `FileId`?

File: src/mito2/src/config.rs (hunk 1 above)
Comment (v0y4g3r) [reply src/mito2/src/config.rs]: > How about putting them in `InvertedIndexConfig`?

fixed in 9b457d930

File: src/mito2/src/cache/index.rs (hunk 3 above)
Comment (v0y4g3r) [reply src/mito2/src/cache/index.rs]: > Can we use `FileId`?

fixed in f1e3558bd17e67e7a3ffe3df1f8cbf5ca9e44ccd

File: src/mito2/src/cache/index.rs (hunk 2 above)
Comment (v0y4g3r) [reply src/mito2/src/cache/index.rs]: > Looks like `put_index()` already added `CACHE_BYTES`.

fixed in f1e3558bd17e67e7a3ffe3df1f8cbf5ca9e44ccd

**Review by zhongzc — APPROVED** [review]

**Review by evenyag — APPROVED** [review]

File: tests-integration/tests/http.rs [hunk 4]
```diff
@@ -839,6 +839,8 @@ create_on_compaction = "auto"
 apply_on_query = "auto"
 mem_threshold_on_create = "auto"
 compress = true
+metadata_cache_size = "32MiB"
+content_cache_size = "32MiB"
```
Comment (killme2008) [inline tests-integration/tests/http.rs:843]: Try to make their default value `auto`.

**Pull request merged**